<a href="https://colab.research.google.com/github/p09323028/2021s_NTU_Econometrics_II/blob/main/Homework/PS1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **經濟實證與預測二_ Homework 1**
**Link:** https://tinyurl.com/vf2894aa

**吳金擇, P09323028**

**林婧筠, P09323026**

**洪靜玟, P09323019**

**林冠伭, P08323019**

# **Exercise 11.8**
Consider the linear probability model $Y_i = \beta_0 + \beta_1 X_i + u_i$, and assume that
$E(u_i|X_i) = 0$.
1. Show that $Pr(Y_i=1|X_i) = \beta_0 + \beta_1X_i$.
2. Show that $Var(u_i|X_i) = \left(\beta_0 + \beta_1 X_i \right) \left[1 - \left( \beta_0 + \beta_1X_i \right) \right]$.
[Hint: Review
Equation (2.7).]
3. Is $u_i$ heteroskedastic? Explain.
4. (Requires Section 11.3) Derive the likelihood function.

---

## 11.8-1

$E(Y_i|X_i) = E(\beta_0 + \beta_1 X_i + u_i|X_i) = \beta_0 + \beta_1 X_i$

若 $Y_i$ 為 binary variable, 

$E(Y_i|X_i) = P(Y_i=0|X_i) \times 0 + P(Y_i=1|X_i) \times 1$

得 $P(Y_i=1|X_i) = \beta_0 + \beta_1 X_i$

## 11.8-2
$Var(u_i|X_i)$

$= (0 - \beta_0 - \beta_1 X_i)^2 \times (1 - \beta_0 - \beta_1 X_i) + (1 - \beta_0 - \beta_1 X_i)^2 \times (\beta_0 + \beta_1 X_i)$

$= (\beta_0 + \beta_1 X_i)[1-(\beta_0 + \beta_1 X_i)]$

## 11.8-3
由 11.8-2 知 $Var(u_i|X_i)$ 隨 $X_i$ 而定， 故 $u_i$ 為 heteroskedastic。

## 11.8-4
由 11.8-1 知 $P(Y_i=1|X_i) = \beta_0 + \beta_1 X_i$,

若 $(Y_i, X_i)$ 為 $\overset{i.i.d.}{\sim}$ $f(Y_i|X_i)$,

$Likelihood$ $Function:$

$Pr(Y_1=y_1, ..., Y_n=y_n|X_1, ..., X_n)$

$ = \prod^n_{i=1}\left[ Pr(Y_i=1|X_i) \right]^{y_i} \left[ Pr(Y_i=0|X_i) \right]^{1-y_i}$

$ = \prod^n_{i=1} (\beta_0 + \beta_1 X_i)^{y_i} (1 - \beta_0 - \beta_1 X_i)^{1-y_i}$

# **Exercise 11.10**
(Requires Section 11.3 and calculus) Suppose a random variable Y has the
following probability distribution: $Pr(Y = 1) = p$, $Pr(Y = 2) = q$, and
$Pr(Y = 3) = 1 - p - q$. A random sample of size n is drawn from this distribution,
and the random variables are denoted $Y_1, Y_2,...,Y_n$.
1. Derive the likelihood function for the parameters $p$ and $q$.
2. Derive formulas for the MLE of $p$ and $q$.

---

## 11.10-1
令樣本中, Y 為 1 的出現次數為 $n_1$, Y 為 2 的出現次數為 $n_2$, Y 為 3 的出現次數為 $n_3 = n - n_1 - n_2$。

$Pr(Y_1=y_1, ..., Y_n=y_n) = \prod^n_{i=1} Pr(Y_i = y_i) = p^{n_1} q^{n_2} (1-p-q)^{n-n_1-n_2}$


## 11.10-2
$L = \text{ln}[Pr(Y_1=y_1, ..., Y_n=y_n)] = n_1 \text{ln}(p) + n_2 \text{ln}(q) + (n-n_1-n_2) \text{ln}(1-p-q)$

$F.O.C.$
- $\frac{\partial{L}}{\partial{p}} = 0 \Rightarrow \frac{n_1}{p} - \frac{n-n_1-n_2}{1-p-q} = 0$ 
- $\frac{\partial{L}}{\partial{q}} = 0 \Rightarrow \frac{n_2}{q} - \frac{n-n_1-n_2}{1-p-q} = 0$ 

得 $\widehat{p}_{MLE} = \frac{n_1}{n}$, $\widehat{q}_{MLE} = \frac{n_2}{n}$

# **Exercise 12.2**
Consider the regression model with a single regressor: $Y_i = \beta_0 + \beta_1 X_i + u_i$. Suppose the least squares assumptions in Key Concept 4.3 are satisfied.
1. Show that $X_i$ is a valid instrument. That is, show that Key Concept 12.3 is
satisfied with $Z_i = X_i$.
2. Show that the IV regression assumptions in Key Concept 12.4 are
satisfied with this choice of $Z_i$.
3. Show that the IV estimator constructed using $Z_i = X_i$ is identical to the
OLS estimator.

---



## 12.2-1
- 由於 $Z_i = X_i$, 故 $Cov(Z_i, X_i) = 1$, 故 $X_i$ 滿足 Instrument Relevance.
- 由於 $X_i$ 滿足 Key Concept 4.3, 知 $Cov(X_i, u_i) = Cov(Z_i, u_i) = 0$, 故 $X_i$ 滿足 Instrument Exogeneity.

## 12.2-2
- 由於無輔助變數$(W_i)$, 且 $X_i$ 滿足 Key Concept 4.3, 故 $E(u_i|W_i) = 0$
- 由 Key Concept 4.3, 知 $(X_i, Z_i, Y_i)$ are i.i.d. draws from their joint distribution.
- 由 Key Concept 4.3, 知 large outliers are unlikely.
- 由 12.2-1 知 $Z_i = X_i$ 符合 Key Concenpt 12.3. 

## 12.2-3
$\widehat{\beta}^{ols}_1 = \frac{S_{xy}}{S_{xx}}$

$\widehat{\beta}^{tsls}_1 = \frac{S_{zy}}{S_{xz}} = \frac{S_{xy}}{S_{xx}}$

$\Rightarrow \widehat{\beta}^{ols}_1 = \widehat{\beta}^{tsls}_1$

# **Exercise 12.5**
Consider the IV regression model
$$Y_i = \beta_0 + \beta_1 X_i + \beta_2 W_i + u_i,$$
where $X_i$ is correlated with $u_i$ and $Z_i$ is an instrument. Suppose that the first three assumptions in Key Concept 12.4 are satisfied. Which IV assumption is
not satisfied when
1. $Z_i$ is independent of $(Y_i, X_i, W_i)$?
2. $Z_i = W_i$?
3. $W_i = 1$ for all $i$?
4. $Z_i = X_i$?

---



## 12.5-1
當 $Z_i$ is independent of $(Y_i, X_i, W_i)$, 則 $Cov(Z_i, X_i) = 0$, 違反 Instrument Relevance。

## 12.5-2
當 $Z_i = W_i$, 在估算第一階迴歸式 (Regress $X_i$ on ($Z_i, W_i$)) 時, 將產生完全線性重合。

## 12.5-3
當 $W_i = 1$, 則將與常數項重合。

## 12.5-4
當 $Z_i = X_i$, 由於 $Cov(X_i, u_i) \neq 0$, 故 $Cov(Z_i, u_i) \neq 0$, 違反 Instrument Exogeneity。

# **Exercise 12.8**
Consider a product market with a supply function $Q^s_i = \beta_0 + \beta_1P_i + u^s_i$, a demand function $Q^d_i = \gamma_0 + u^d_i$, and a market equilibrium condition
$Q^s_i = Q^d_i$, where $u^s_i$ and $u^d_i$
are mutually independent $i.i.d$. random variables, both with a mean of $0$.
1. Show that $P_i$ and $u^s_i$ are correlated.
2. Show that the OLS estimator of $\beta_1$ is inconsistent.
3. How would you estimate $\beta_0$, $\beta_1$, and $\gamma_0$?

---



## 12.8-1
$\beta_0 + \beta_1 P_i + u_i^s = \gamma_0 + u_i^d$

$P_i = \frac{\gamma_0 - \beta_0 + u_i^d - u_i^s}{\beta_1}$

$\Rightarrow Cov(P_i, u_i^s) = \frac{-\sigma_{u^s}}{\beta_1}$


## 12.8-2
因 $Cov(P_i, u_i^s) \neq 0$, 故 $\widehat{\beta}_1^{ols}$ is inconsistent。

## 12.8-3
此例中，需求函數完全無彈性($\varepsilon^d = \frac{\partial{Q^d}}{\partial{P}} \frac{P}{Q} = 0$),

知 $Cov(Q_i^d, u_i^s) = 0$ 且 $Cov(Q_i^d, P_i) \neq 0$, 

故可將 $Q^d$ 視為工具變數, 採 TSLS 估算 $\beta_0$ 及 $\beta_1$,

而 $\gamma_0$ 則可直接採 OLS 估算。 

# **Exercise 12.10**
Two classmates are comparing their answers to an assignment. One classmate has specified an instrumental variable regression model $Y_i = \beta_0 + \beta_1 X_i + \beta_2 W_i + u_i$, using $Z_i$ as an instrument. The other student has specified the same model, but has omitted $W_i$.
1. The first student says that if $Z_i$ and $W_i$ are correlated, then the second
student’s IV estimator will not be consistent. Is the first student right
about this?
2. The second student argues that if in the true model $\beta_2 = 0$, then their IV
estimator will be consistent. Is the second student correct in saying this?

---



## 12.10-1
$ \beta_1 = \frac{Cov(Z_i,Y_i)}{Cov(Z_i,X_i)} = \frac{Cov(Z_i, \beta_0 + \beta_1 X_i + u_i^{'})}{Cov(Z_i, X_i)} $

$ = \frac{\beta_1 Cov(Z_i, X_i) + Cov(Z_i, u^{'}_i)}{Cov(Z_i, X_i)} = \frac{\beta_1 Cov(Z_i, X_i) + Cov(Z_i, \beta_2 W_i + u_i)}{Cov(Z_i, X_i)} $

由上式知，$Cov(Z_i, W_i) \neq 0 \Rightarrow Cov(Z_i, u^{'}_i) \neq 0$,

違反 Instrument Exogeneity,

第二位學生的估計式 $\widehat{\beta}^{TSLS}_1$ 非為一致估計式。


## 12.10-2
$ \beta_1 = \frac{Cov(Z_i,Y_i)}{Cov(Z_i,X_i)} = \frac{Cov(Z_i, \beta_0 + \beta_1 X_i + u_i^{'})}{Cov(Z_i, X_i)} $

$= \frac{\beta_1 Cov(Z_i, X_i) + Cov(Z_i, u^{'}_i)}{Cov(Z_i, X_i)} = \frac{\beta_1 Cov(Z_i, X_i) + Cov(Z_i, \beta_2 W_i + u_i)}{Cov(Z_i, X_i)} $

若未能確保 $Cov(X_i, u^{'}_i) =0$, 即使 $\beta_2 = 0$, 第二位學生的估計式 $\widehat{\beta}^{TSLS}_1$ 仍非為一致估計式。

# **E12.2**
Does viewing a violent movie lead to violent behavior? If so, the incidence of
violent crimes, such as assaults, should rise following the release of a violent
movie that attracts many viewers. Alternatively, movie viewing may substitute
for other activities (such as alcohol consumption) that lead to violent behavior, so that assaults should fall when more viewers are attracted to the cinema. On the text website, **http://www.pearsonglobaleditions.com**, you will find the data file **Movies**, which contains data on the number of assaults and movie attendance for 516 weekends from 1995 through 2004.7 A detailed description is given in **Movies_Description**, available on the website. The data set includes weekend U.S. attendance for strongly violent movies (such as Hannibal), mildly violent movies (such as Spider-Man), and nonviolent movies (such as Finding Nemo). The data set also includes a count of the number of assaults for the same weekend in a subset of counties in the United States. Finally, the data set includes indicators for year, month, whether the weekend is a holiday, and various measures of the weather.
1. - Regress the logarithm of the number of assaults $[ln\_assaults=\text{ln}
(assaults)]$ on the year and month indicators. Is there evidence of
seasonality in assaults? That is, do there tend to be more assaults in
some months than others? Explain.
 - Regress total movie attendance $(attend = attend\_v + attend\_m +
attend\_n)$ on the year and month indicators. Is there evidence of
seasonality in movie attendance? Explain.

2. Regress $ln\_assaults$ on $attend\_v, attend\_m, attend\_n$, the year and month
indicators, and the weather and holiday control variables available in the
data set.
 - Based on the regression, does viewing a strongly violent movie increase or decrease assaults? By how much? Is the estimated effect statistically significant?
 - Does attendance at strongly violent movies affect assaults differently than attendance at moderately violent movies? Differently than attendance at nonviolent movies?
 - A strongly violent blockbuster movie is released, and the weekend’s attendance at strongly violent movies increases by 6 million; meanwhile, attendance falls by 2 million for moderately violent movies
and by 1 million for nonviolent movies. What is the predicted effect on assaults? Construct a 95% confidence interval for the change in assaults. [Hint: Review Section 7.3 and material surrounding Equations (8.7) and (8.8).]
3. It is difficult to control for all the variables that affect assaults and that might be correlated with movie attendance. For example, the effect of
the weather on assaults and movie attendance is only crudely approximated
by the weather variables in the data set. However, the data set does include a set of instruments—$pr\_attend\_v, pr\_attend\_m$, and $pr\_attend\_n$—that are correlated with attendance but are (arguably) uncorrelated with weekend-specific factors (such as the weather) that affect both assaults and movie attendance. These instruments use historical attendance patterns, not information on a particular weekend, to predict a film’s attendance in a given weekend. For example, if a film’s attendance is high in the second week of its release, then this can be used to predict that its attendance was also high in the first week of its release. (The details of the construction of these instruments are available in the Dahl and DellaVigna paper referenced in footnote 5.) Run the regression from (b) (including year, month, holiday, and weather controls) but now using $pr\_attend\_v, pr\_attend\_m$, and $pr\_attend\_n$ as instruments for $attend\_v, attend\_m$, and $attend\_n$. Use this IV regression to answer (b)(i)–(b)(iii).
4. The intuition underlying the instruments in (c) is that attendance in
a given week is correlated with attendance in surrounding weeks. For each movie category, the data set includes attendance in surrounding weeks. Run the regression using the instruments $attend\_v\_f, attend\_m\_f, attend\_n\_f, attend\_v\_b, attend\_m\_b,$ and $attend\_n\_b$ instead of the instruments used in (c). Use this IV regression to answer (b)(i)–(b)(iii).
5. There are nine instruments listed in (c) and (d), but only three are
needed for identification. Carry out the test for overidentification summarized
in Key Concept 12.6. What do you conclude about the validity of the instruments?
6. Based on your analysis, what do you conclude about the effect of violent
movies on (short-run) violent behavior?

---



下載資料

In [ ]:
!gdown --id '1IQZjS8Mqpqf3R-rd7V4kMvilXI0Tqj42' --output movie.xlsx

下載及引入套件

In [ ]:
!pip install linearmodels
import numpy as np
import pandas as pd
from statsmodels.sandbox.regression.gmm import IV2SLS
import statsmodels.formula.api as smf
import linearmodels.iv as iv

載入並檢視資料

In [3]:
movie = pd.read_excel('movie.xlsx')

In [ ]:
movie.columns

設置變數

In [5]:
movie['ln_assaults'] = np.log(movie['assaults']) 
movie['attend'] = movie['attend_v'] + movie['attend_m'] + movie['attend_n']
movie['constant'] = 1

In [ ]:
movie.head()

## E12.2-a

### E12.2-a(i)

In [7]:
reg_ai = smf.ols(formula='ln_assaults~year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12', data=movie)
results_ai = reg_ai.fit(cov_type='HC1')
print(results_ai.summary())

f_test = results_ai.f_test(['month2=0','month3=0','month4=0','month5=0','month6=0','month7=0','month8=0','month9=0','month10=0','month11=0','month12=0'])
f_test.pvalue

                            OLS Regression Results                            
Dep. Variable:            ln_assaults   R-squared:                       0.992
Model:                            OLS   Adj. R-squared:                  0.991
Method:                 Least Squares   F-statistic:                     3081.
Date:                Fri, 09 Apr 2021   Prob (F-statistic):               0.00
Time:                        07:36:14   Log-Likelihood:                 741.74
No. Observations:                 516   AIC:                            -1441.
Df Residuals:                     495   BIC:                            -1352.
Df Model:                          20                                         
Covariance Type:                  HC1                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.7276      0.016    423.870      0.0

array(7.3525147e-101)

$\alpha = 0.1, Pvalue=0.00 < \alpha \Rightarrow$ $ln\_assaults$ 與 $month$ 有顯著關係。(4月~10月偏高)

### E12.2-a(ii)

In [8]:
reg_aii = smf.ols(formula='attend~year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12', data=movie)
results_aii = reg_aii.fit(cov_type='HC1')
print(results_aii.summary())

f_test = results_aii.f_test(['month2=0','month3=0','month4=0','month5=0','month6=0','month7=0','month8=0','month9=0','month10=0','month11=0','month12=0'])
f_test.pvalue

                            OLS Regression Results                            
Dep. Variable:                 attend   R-squared:                       0.480
Model:                            OLS   Adj. R-squared:                  0.459
Method:                 Least Squares   F-statistic:                     42.91
Date:                Fri, 09 Apr 2021   Prob (F-statistic):           1.56e-94
Time:                        07:36:14   Log-Likelihood:                -1383.6
No. Observations:                 516   AIC:                             2809.
Df Residuals:                     495   BIC:                             2898.
Df Model:                          20                                         
Covariance Type:                  HC1                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     16.0396      0.746     21.492      0.0

array(2.32251358e-82)

$\alpha = 0.1, Pvalue=0.00 < \alpha \Rightarrow$ $attend$ 與 $month$ 有顯著關係。(6、7、8、11月偏高)

## E12.2-b

### E12.2-b(i)

In [9]:
reg_bi = smf.ols(formula='ln_assaults~attend_v+attend_m+attend_n+year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12+h_chris+h_newyr+h_easter+h_july4+h_mem+h_labor+w_rain+w_snow+w_maxa+w_maxb+w_maxc+w_mina+w_minb+w_minc', data=movie)
results_bi = reg_bi.fit(cov_type='HC1')
print(results_bi.summary())

                            OLS Regression Results                            
Dep. Variable:            ln_assaults   R-squared:                       0.996
Model:                            OLS   Adj. R-squared:                  0.996
Method:                 Least Squares   F-statistic:                     3166.
Date:                Fri, 09 Apr 2021   Prob (F-statistic):               0.00
Time:                        07:36:14   Log-Likelihood:                 924.66
No. Observations:                 516   AIC:                            -1773.
Df Residuals:                     478   BIC:                            -1612.
Df Model:                          37                                         
Covariance Type:                  HC1                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.9143      0.020    346.987      0.0

Coef. of $attend\_v$: $-0.32%$, $attend\_v$ 增加一單位, $ln\_assaults$ 將 $-0.32\%$

$\alpha = 0.1, Pvalue=0.00 < \alpha \Rightarrow$ $attend\_v$ 對 $ln\_assaults$ 的影響顯著異於零

### E12.2-b(ii)

In [10]:
f_test = results_bi.f_test(['attend_v=attend_m'])
print(f_test)

<F test: F=array([[0.00143182]]), p=0.9698315067888446, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue=0.97 > \alpha \Rightarrow$ 兩者無顯著差異

In [11]:
f_test = results_bi.f_test(['attend_v=attend_n'])
print(f_test)

<F test: F=array([[1.63178954]]), p=0.20207580002483425, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue=0.20 > \alpha \Rightarrow$ 兩者無顯著差異

### E12.2-b(iii)

In [12]:
t_test = results_bi.t_test(['6*attend_v-2*attend_m-1*attend_n=0'])
t_test.summary(xname=['attend'], alpha=0.05, title='Effect on Assaults')

<class 'statsmodels.iolib.table.SimpleTable'>

Predicted effect on assaults: $-1.06\%$

$95\%$ 信賴區間: $\left[ -2\%, -0.1\% \right]$

## E12.2-c

### E12.2-c(i)

In [13]:
reg_ci = iv.IV2SLS.from_formula(formula='ln_assaults~1+[attend_v+attend_m+attend_n~pr_attend_v+pr_attend_m+pr_attend_n]+year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12+h_chris+h_newyr+h_easter+h_july4+h_mem+h_labor+w_rain+w_snow+w_maxa+w_maxb+w_maxc+w_mina+w_minb+w_minc', data=movie)
results_ci = reg_ci.fit(cov_type='robust', debiased=True)
print(results_ci.summary)

                          IV-2SLS Estimation Summary                          
Dep. Variable:            ln_assaults   R-squared:                      0.9959
Estimator:                    IV-2SLS   Adj. R-squared:                 0.9956
No. Observations:                 516   F-statistic:                    3121.8
Date:                Fri, Apr 09 2021   P-value (F-stat)                0.0000
Time:                        07:36:14   Distribution:                F(37,478)
Cov. Estimator:                robust                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      6.9242     0.0187     370.24     0.0000      6.8874      6.9609
year2          0.7019     0.0098     71.630     0.00

Coef. of $attend\_v$: $-0.39\%$, $attend\_v$ 增加一單位, $ln\_assaults$ 將 $-0.39\%$

$\alpha = 0.1, Pvalue= 0.00 < \alpha \Rightarrow$ $attend\_v$ 對 $ln\_assaults$ 的影響顯著異於零

### E12.2-c(ii)

In [14]:
# iv.IV2SLS 好似無 partial F test
# 改用 Statsmodels....IV2SLS, 但似乎無法調整 robust
endog = movie['ln_assaults']
exog = movie[['constant','attend_v','attend_m','attend_n','year2','year3','year4','year5','year6','year7','year8','year9','year10','month2','month3','month4','month5','month6','month7','month8','month9','month10','month11','month12','h_chris','h_newyr','h_easter','h_july4','h_mem','h_labor','w_maxa','w_maxb','w_maxc','w_mina','w_minb','w_minc','w_rain','w_snow']]
instrument = movie[['constant','pr_attend_v','pr_attend_m','pr_attend_n','year2','year3','year4','year5','year6','year7','year8','year9','year10','month2','month3','month4','month5','month6','month7','month8','month9','month10','month11','month12','h_chris','h_newyr','h_easter','h_july4','h_mem','h_labor','w_maxa','w_maxb','w_maxc','w_mina','w_minb','w_minc','w_rain','w_snow']]

results_ci = IV2SLS(endog=endog,exog=exog,instrument=instrument).fit()  # 似乎尚無法調整 robust ...

In [15]:
f_test = results_ci.f_test(['attend_v=attend_m'])
print(f_test)

<F test: F=array([[0.00041368]]), p=0.9837812840120339, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue=0.98 > \alpha \Rightarrow$ 兩者統計上無顯著差異。

In [16]:
f_test = results_ci.f_test(['attend_v=attend_n'])
print(f_test)

<F test: F=array([[1.34367722]]), p=0.24696525542942116, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue=0.25 > \alpha \Rightarrow$ 兩者統計上無顯著差異。

### E12.2-c(iii)

In [17]:
t_test = results_ci.t_test(['6*attend_v-2*attend_m-1*attend_n=0'])
t_test.summary(xname=['attend'], alpha=0.05, title='Effect on Assaults')

<class 'statsmodels.iolib.table.SimpleTable'>

Predicted effect on assaults: $-1.27\%$

$95\%$ 信賴區間: $\left[ -2.4\%, -0.2\% \right]$

## E12.2-d

### E12.2-d(i)

In [18]:
reg_di = iv.IV2SLS.from_formula(formula='ln_assaults~1+[attend_v+attend_m+attend_n~attend_v_f+attend_m_f+attend_n_f+attend_v_b+attend_m_b+attend_n_b]+year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12+h_chris+h_newyr+h_easter+h_july4+h_mem+h_labor+w_rain+w_snow+w_maxa+w_maxb+w_maxc+w_mina+w_minb+w_minc', data=movie)
results_di = reg_di.fit(cov_type='robust', debiased=True)
print(results_di.summary)

                          IV-2SLS Estimation Summary                          
Dep. Variable:            ln_assaults   R-squared:                      0.9959
Estimator:                    IV-2SLS   Adj. R-squared:                 0.9956
No. Observations:                 516   F-statistic:                    3113.3
Date:                Fri, Apr 09 2021   P-value (F-stat)                0.0000
Time:                        07:36:14   Distribution:                F(37,478)
Cov. Estimator:                robust                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      6.9225     0.0291     238.04     0.0000      6.8654      6.9797
year2          0.7025     0.0098     71.932     0.00

Coef. of $attend\_v$: $-0.32\%$, $attend\_v$ 增加一單位, $ln\_assaults$ 將 $-0.32\%$

$\alpha = 0.1, Pvalue= 0.1037 > \alpha \Rightarrow$ $attend\_v$ 對 $ln\_assaults$ 無顯著影響

### E12.2-d(ii)

In [19]:
# iv.IV2SLS 好似無 partial F test
# 改用 Statsmodels....IV2SLS, 但似乎無法調整 robust
exog = movie[['constant','attend_v','attend_m','attend_n','year2','year3','year4','year5','year6','year7','year8','year9','year10','month2','month3','month4','month5','month6','month7','month8','month9','month10','month11','month12','h_chris','h_newyr','h_easter','h_july4','h_mem','h_labor','w_maxa','w_maxb','w_maxc','w_mina','w_minb','w_minc','w_rain','w_snow']]
instrument = movie[['constant','attend_v_f','attend_m_f','attend_n_f','attend_v_b','attend_m_b','attend_n_b','year2','year3','year4','year5','year6','year7','year8','year9','year10','month2','month3','month4','month5','month6','month7','month8','month9','month10','month11','month12','h_chris','h_newyr','h_easter','h_july4','h_mem','h_labor','w_maxa','w_maxb','w_maxc','w_mina','w_minb','w_minc','w_rain','w_snow']]

results_di = IV2SLS(endog=endog,exog=exog,instrument=instrument).fit()

In [20]:
f_test = results_di.f_test(['attend_v=attend_m'])
print(f_test)

<F test: F=array([[0.37994427]]), p=0.5379264960024015, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue= 0.54 > \alpha \Rightarrow$ 兩者統計上無顯著差異。

In [21]:
f_test = results_di.f_test(['attend_v=attend_n'])
print(f_test)

<F test: F=array([[0.17517543]]), p=0.6757405981234799, df_denom=478, df_num=1>


$\alpha = 0.1, Pvalue= 0.68 > \alpha \Rightarrow$ 兩者統計上無顯著差異。

### E12.2-d(iii)

In [22]:
t_test = results_di.t_test(['6*attend_v-2*attend_m-1*attend_n=0'])
t_test.summary(xname=['attend'], alpha=0.05, title='Effect on Assaults')

<class 'statsmodels.iolib.table.SimpleTable'>

Predicted effect on assaults: $-0.82\%$

$95\%$ 信賴區間: $\left[ -2.7\%, 1.1\% \right]$

## E12.2-e

In [23]:
reg_e = iv.IV2SLS.from_formula(formula='ln_assaults~1+[attend_v+attend_m+attend_n~pr_attend_v+pr_attend_m+pr_attend_n+attend_v_f+attend_m_f+attend_n_f+attend_v_b+attend_m_b+attend_n_b]+year2+year3+year4+year5+year6+year7+year8+year9+year10+month2+month3+month4+month5+month6+month7+month8+month9+month10+month11+month12+h_chris+h_newyr+h_easter+h_july4+h_mem+h_labor+w_rain+w_snow+w_maxa+w_maxb+w_maxc+w_mina+w_minb+w_minc', data=movie)
results_e = reg_e.fit(cov_type='robust')
movie['e'] = results_e.resids

j_test = smf.ols(formula='e~pr_attend_v+pr_attend_m+pr_attend_n+attend_v_f+attend_m_f+attend_n_f+attend_v_b+attend_m_b+attend_n_b', data=movie).fit()
print('Overid. stat: {:.2f}'.format(j_test.fvalue*j_test.df_model))

Overid. stat: 8.11


j-stat: 8.11

自由度為 $9 - 3 = 6$ 的卡方分配 $\chi^2_{0.1}(6) = 10.645$,

故無證據顯示這些工具變數違反Instrument Exogeneity。

## E12.2-f

由上述分析可知，短期內, 看電影與暴力行為(如喝酒肇事)兩者互相替代，故無論看何種類型的電影，都可減少暴力事件發生。